R/table-ident.R

```diff
@@ -0,0 +1,244 @@
+                            error_call = caller_env()) {
+  check_dots_empty(call = error_call)
+
+  check_character(table, call = error_call)
+  check_character(schema, call = error_call)
+  check_character(catalog, call = error_call)
+  check_logical(quoted, call = error_call)
+
+  n <- vctrs::vec_size_common(
+    table = table,
+    schema = schema,
+    catalog = catalog,
+    quoted = quoted,
+    .call = error_call
+  )
+  data <- list(
+    table = vctrs::vec_recycle(table, n, call = error_call),
+    schema = vctrs::vec_recycle(schema, n, call = error_call),
+    catalog = vctrs::vec_recycle(catalog, n, call = error_call),
+    quoted = vctrs::vec_recycle(quoted, n, call = error_call),
+    name = vctrs::vec_rep(NA_character_, n)
+  )
+
+  purrr::pmap(
+    data,
+    function(table, schema, catalog, quoted, name) {
+      check_db_table_args(
+        table = table,
+        schema = schema,
+        catalog = catalog,
+        quoted = quoted,
+        call = error_call
+      )
+    }
+  )
+
+  vctrs::new_rcrd(data, class = "dbplyr_table_ident")
+}
+
+check_db_table_args <- function(table,
+                                schema,
+                                catalog,
+                                quoted,
+                                call = caller_env()) {
+  if (!is.na(catalog) && is.na(schema)) {
+    cli_abort("Must supply {.arg schema} when {.arg catalog} is supplied.", call = call)
+  }
+  if (!is.na(schema) && is.na(table)) {
+    cli_abort("Must supply {.arg table} when {.arg schema} is supplied.", call = call)
+  }
+
+  if (quoted && !is.na(schema)) {
+    cli_abort("Can't supply a schema when {.arg table} is quoted.", call = call)
+  }
+}
+
+as_table_ident <- function(x, ..., error_call = current_env()) {
+  check_dots_empty()
+  UseMethod("as_table_ident")
+}
+
+#' @export
+as_table_ident.dbplyr_table_ident <- function(x, ..., error_call = current_env()) {
+  x
+}
+
+#' @export
+as_table_ident.character <- function(x, ..., error_call = current_env()) {
+  new_table_ident(table = vctrs::vec_data(x))
+}
+
+#' @export
+as_table_ident.ident <- function(x, ..., error_call = current_env()) {
+  new_table_ident(table = vctrs::vec_data(x))
+}
+
+#' @export
+as_table_ident.ident_q <- function(x, ..., error_call = current_env()) {
+  inform_unwanted_table_ident("ident_q")
+  new_table_ident(table = vctrs::vec_data(x), quoted = TRUE)
+}
+
+#' @export
+as_table_ident.sql <- function(x, ..., error_call = current_env()) {
+  inform_unwanted_table_ident("sql")
+  new_table_ident(table = vctrs::vec_data(x), quoted = TRUE)
+}
+
+inform_unwanted_table_ident <- function(f) {
+  cli::cli_inform(
+    c(
+      i = "Using {.fn {f}} for a table identifier is intended as fallback in case of bugs.",
+      i = "If you need it to work around a bug please open an issue {.url https://github.com/tidyverse/dbplyr/issues}."
+    ),
+    .frequency = "regularly",
+    .frequency_id = paste0("dbplyr_", f)
+  )
```

Comment:
This should hopefully push people to avoid using `ident_q()` and `sql()` unless necessary.